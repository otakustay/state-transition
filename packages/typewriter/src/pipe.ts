import {IterationItem, Queue, QueueState} from './queue.js';

async function consumeInput(input: AsyncIterableIterator<string>, queue: Queue) {
    const cursor = {current: 0};
    for await (const chunk of input) {
        queue.put(cursor.current, chunk);
        cursor.current++;
    }
    queue.complete();
}

interface TypewriterStrategyContext {
    getQueueState: () => QueueState;
}

type TypewriterStrategy = (chunk: IterationItem, context: TypewriterStrategyContext) => AsyncIterableIterator<string>;

interface TypewriterOptions {
    strategy: TypewriterStrategy;
}

function wait(time: number): Promise<void> {
    if (time <= 0) {
        return Promise.resolve();
    }

    return new Promise(resolve => setTimeout(resolve, time));
}

export function byChunk(): TypewriterStrategy {
    return async function* byChunkStrategy(chunk) {
        yield chunk.value;
    };
}

export function byCharacterLinear(interval: number): TypewriterStrategy {
    return async function* byCharacterLinearStrategy(chunk) {
        for (const char of chunk.value) {
            yield char;
            await wait(interval);
        }
    };
}

export function byWordLinear(interval: number): TypewriterStrategy {
    return async function* byWordLinearStrategy(chunk) {
        const segmenter = new Intl.Segmenter('zh-Hans', {granularity: 'word'});
        const segments = segmenter.segment(chunk.value);
        for (const data of segments) {
            yield data.segment;
            await wait(interval);
        }
    };
}

export interface EagerOptions {
    defaultInterval: number;
    eagerInterval: number;
}

export function byWordEager({defaultInterval, eagerInterval}: EagerOptions): TypewriterStrategy {
    return async function* byWordEagerStrategy(chunk, context) {
        const segmenter = new Intl.Segmenter('zh-Hans', {granularity: 'word'});
        const segments = segmenter.segment(chunk.value);
        for (const data of segments) {
            const state = context.getQueueState();
            const remainingChunksCount = state.resolved.length - chunk.index - 1;

            // Once we are in a state that too behind of incoming stream, we should flush all possible text immediately
            if (state.completed || remainingChunksCount >= 2) {
                yield chunk.value.slice(data.index);
                return;
            }

            // For condition where not that much backpressure, we can still output word by word in a smaller interval
            const interval = remainingChunksCount === 1 ? eagerInterval : defaultInterval;
            yield data.segment;
            await wait(interval);
        }
    };
}

export async function* createTypewriterPipeline(input: AsyncIterableIterator<string>, options: TypewriterOptions) {
    const queue = new Queue();
    void consumeInput(input, queue);
    const cursor = {index: 0};

    for await (const chunk of queue.toIterable()) {
        const context: TypewriterStrategyContext = {
            getQueueState: () => queue.getState(),
        };
        for await (const output of options.strategy(chunk, context)) {
            yield output;
        }

        cursor.index++;
    }
}
