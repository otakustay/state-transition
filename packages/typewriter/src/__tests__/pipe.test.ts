import {test, expect, vi} from 'vitest';
import {byCharacterLinear, createTypewriterPipeline} from '../pipe.js';
import {byChunk, byWordEager, byWordLinear} from '../index.js';

vi.setConfig({testTimeout: 30000});

const textChunks = [
    '晴朗的天空',
    '下，阳光如金色绸带般洒落，空气中弥漫着花草的香气。',
    '在这和煦的气氛中，一只可爱的小猫正在草丛中欢快地玩耍。',
    '它的毛色如同雪一般洁白，间或夹杂着几缕浅浅的灰色。',
    '小猫灵动活泼，不时地在草丛中跳跃翻滚，好似在与风儿捉迷藏。',
    '它的蓝色双眼闪烁着好奇与探索的光芒，似乎对这周围的一切都充满了无尽的兴趣。',
    '草丛中的小虫、蝴蝶都成了它的玩伴，它时而扑腾着追逐，时而安静地观察，一举一动都透露着童真与可爱',
    '。',
    '偶尔，它会停下来，昂首望向天空，仿佛在对着那洁白的云朵诉说着自己的小心事。',
    '阳光照耀在小猫柔软的毛发上，形成一层淡淡的光晕，这一刻，它就像是草丛中的小精灵，为这美好的天气增添了一抹生动的',
    '色彩。',
];

async function* incoming(latency = 10): AsyncIterableIterator<string> {
    for (const chunk of textChunks) {
        await new Promise(resolve => setTimeout(resolve, latency));
        yield chunk;
    }
}

async function streamToArray(stream: AsyncIterableIterator<string>) {
    const result: string[] = [];
    for await (const chunk of stream) {
        result.push(chunk);
    }
    return result;
}

test('by chunk', async () => {
    const pipe = createTypewriterPipeline(
        incoming(),
        {
            strategy: byChunk(),
        }
    );
    const result = await streamToArray(pipe);
    expect(result).toEqual(textChunks);
});

test('character linear', async () => {
    const pipe = createTypewriterPipeline(
        incoming(),
        {
            strategy: byCharacterLinear(0),
        }
    );
    const result = await streamToArray(pipe);
    expect(result).toEqual(textChunks.map(v => v.split('')).flat());
});

test('word linear', async () => {
    const pipe = createTypewriterPipeline(
        incoming(),
        {
            strategy: byWordLinear(0),
        }
    );
    const result = await streamToArray(pipe);
    expect(result.join('')).toBe(textChunks.join(''));
});

test('word eager yield all chunk on pending', async () => {
    const pipe = createTypewriterPipeline(
        incoming(),
        {
            strategy: byWordEager({defaultInterval: 40, eagerInterval: 10}),
        }
    );
    const result = await streamToArray(pipe);
    expect(result.some(v => v.length > 10)).toBe(true);
});

test('word eager await latency on last chunk', async () => {
    const pipe = createTypewriterPipeline(
        incoming(40),
        {
            strategy: byWordEager({defaultInterval: 20, eagerInterval: 10}),
        }
    );
    const result = await streamToArray(pipe);
    expect(result.filter(v => v.length === 1).length).toBeGreaterThan(10);
});

test('error', async () => {
    async function* incoming() {
        yield 'foo';
        throw new Error('bar');
    }

    const pipe = createTypewriterPipeline(
        incoming(),
        {
            strategy: byChunk(),
        }
    );
    await expect(streamToArray(pipe)).rejects.toThrow();
});
