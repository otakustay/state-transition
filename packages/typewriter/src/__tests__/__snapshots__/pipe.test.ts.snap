// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`by chunk 1`] = `
[
  "晴朗的天空",
  "下，阳光如金色绸带般洒落，空气中弥漫着花草的香气。",
  "在这和煦的气氛中，一只可爱的小猫正在草丛中欢快地玩耍。",
  "它的毛色如同雪一般洁白，间或夹杂着几缕浅浅的灰色。",
  "小猫灵动活泼，不时地在草丛中跳跃翻滚，好似在与风儿捉迷藏。",
  "它的蓝色双眼闪烁着好奇与探索的光芒，似乎对这周围的一切都充满了无尽的兴趣。",
  "草丛中的小虫、蝴蝶都成了它的玩伴，它时而扑腾着追逐，时而安静地观察，一举一动都透露着童真与可爱",
  "。",
  "偶尔，它会停下来，昂首望向天空，仿佛在对着那洁白的云朵诉说着自己的小心事。",
  "阳光照耀在小猫柔软的毛发上，形成一层淡淡的光晕，这一刻，它就像是草丛中的小精灵，为这美好的天气增添了一抹生动的",
  "色彩。",
]
`;

exports[`character linear 1`] = `
[
  "晴",
  "朗",
  "的",
  "天",
  "空",
  "下",
  "，",
  "阳",
  "光",
  "如",
  "金",
  "色",
  "绸",
  "带",
  "般",
  "洒",
  "落",
  "，",
  "空",
  "气",
  "中",
  "弥",
  "漫",
  "着",
  "花",
  "草",
  "的",
  "香",
  "气",
  "。",
  "在",
  "这",
  "和",
  "煦",
  "的",
  "气",
  "氛",
  "中",
  "，",
  "一",
  "只",
  "可",
  "爱",
  "的",
  "小",
  "猫",
  "正",
  "在",
  "草",
  "丛",
  "中",
  "欢",
  "快",
  "地",
  "玩",
  "耍",
  "。",
  "它",
  "的",
  "毛",
  "色",
  "如",
  "同",
  "雪",
  "一",
  "般",
  "洁",
  "白",
  "，",
  "间",
  "或",
  "夹",
  "杂",
  "着",
  "几",
  "缕",
  "浅",
  "浅",
  "的",
  "灰",
  "色",
  "。",
  "小",
  "猫",
  "灵",
  "动",
  "活",
  "泼",
  "，",
  "不",
  "时",
  "地",
  "在",
  "草",
  "丛",
  "中",
  "跳",
  "跃",
  "翻",
  "滚",
  "，",
  "好",
  "似",
  "在",
  "与",
  "风",
  "儿",
  "捉",
  "迷",
  "藏",
  "。",
  "它",
  "的",
  "蓝",
  "色",
  "双",
  "眼",
  "闪",
  "烁",
  "着",
  "好",
  "奇",
  "与",
  "探",
  "索",
  "的",
  "光",
  "芒",
  "，",
  "似",
  "乎",
  "对",
  "这",
  "周",
  "围",
  "的",
  "一",
  "切",
  "都",
  "充",
  "满",
  "了",
  "无",
  "尽",
  "的",
  "兴",
  "趣",
  "。",
  "草",
  "丛",
  "中",
  "的",
  "小",
  "虫",
  "、",
  "蝴",
  "蝶",
  "都",
  "成",
  "了",
  "它",
  "的",
  "玩",
  "伴",
  "，",
  "它",
  "时",
  "而",
  "扑",
  "腾",
  "着",
  "追",
  "逐",
  "，",
  "时",
  "而",
  "安",
  "静",
  "地",
  "观",
  "察",
  "，",
  "一",
  "举",
  "一",
  "动",
  "都",
  "透",
  "露",
  "着",
  "童",
  "真",
  "与",
  "可",
  "爱",
  "。",
  "偶",
  "尔",
  "，",
  "它",
  "会",
  "停",
  "下",
  "来",
  "，",
  "昂",
  "首",
  "望",
  "向",
  "天",
  "空",
  "，",
  "仿",
  "佛",
  "在",
  "对",
  "着",
  "那",
  "洁",
  "白",
  "的",
  "云",
  "朵",
  "诉",
  "说",
  "着",
  "自",
  "己",
  "的",
  "小",
  "心",
  "事",
  "。",
  "阳",
  "光",
  "照",
  "耀",
  "在",
  "小",
  "猫",
  "柔",
  "软",
  "的",
  "毛",
  "发",
  "上",
  "，",
  "形",
  "成",
  "一",
  "层",
  "淡",
  "淡",
  "的",
  "光",
  "晕",
  "，",
  "这",
  "一",
  "刻",
  "，",
  "它",
  "就",
  "像",
  "是",
  "草",
  "丛",
  "中",
  "的",
  "小",
  "精",
  "灵",
  "，",
  "为",
  "这",
  "美",
  "好",
  "的",
  "天",
  "气",
  "增",
  "添",
  "了",
  "一",
  "抹",
  "生",
  "动",
  "的",
  "色",
  "彩",
  "。",
]
`;

exports[`word eager 1`] = `
[
  "晴朗",
  "的天空",
  "下，阳光如金色绸带般洒落，空气中弥漫着花草的香气。",
  "在",
  "这和煦的气氛中，一只可爱的小猫正在草丛中欢快地玩耍。",
  "它的",
  "毛色如同雪一般洁白，间或夹杂着几缕浅浅的灰色。",
  "小猫",
  "灵动活泼，不时地在草丛中跳跃翻滚，好似在与风儿捉迷藏。",
  "它的",
  "蓝色双眼闪烁着好奇与探索的光芒，似乎对这周围的一切都充满了无尽的兴趣。",
  "草丛",
  "中的小虫、蝴蝶都成了它的玩伴，它时而扑腾着追逐，时而安静地观察，一举一动都透露着童真与可爱",
  "。",
  "偶尔",
  "，它会停下来，昂首望向天空，仿佛在对着那洁白的云朵诉说着自己的小心事。",
  "阳光照耀在小猫柔软的毛发上，形成一层淡淡的光晕，这一刻，它就像是草丛中的小精灵，为这美好的天气增添了一抹生动的",
  "色彩。",
]
`;

exports[`word linear 1`] = `
[
  "晴朗",
  "的",
  "天空",
  "下",
  "，",
  "阳光",
  "如",
  "金色",
  "绸",
  "带",
  "般",
  "洒落",
  "，",
  "空气",
  "中",
  "弥漫着",
  "花草",
  "的",
  "香气",
  "。",
  "在",
  "这",
  "和煦",
  "的",
  "气氛",
  "中",
  "，",
  "一只",
  "可爱",
  "的",
  "小猫",
  "正在",
  "草丛",
  "中",
  "欢",
  "快",
  "地",
  "玩耍",
  "。",
  "它的",
  "毛色",
  "如同",
  "雪",
  "一般",
  "洁白",
  "，",
  "间",
  "或",
  "夹杂着",
  "几缕",
  "浅",
  "浅",
  "的",
  "灰色",
  "。",
  "小猫",
  "灵动",
  "活泼",
  "，",
  "不时",
  "地",
  "在",
  "草丛",
  "中",
  "跳跃",
  "翻滚",
  "，",
  "好似",
  "在",
  "与",
  "风",
  "儿",
  "捉迷藏",
  "。",
  "它的",
  "蓝色",
  "双眼",
  "闪烁",
  "着",
  "好奇",
  "与",
  "探索",
  "的",
  "光芒",
  "，",
  "似乎",
  "对",
  "这",
  "周围",
  "的",
  "一切都",
  "充满了",
  "无",
  "尽",
  "的",
  "兴趣",
  "。",
  "草丛",
  "中的",
  "小虫",
  "、",
  "蝴蝶",
  "都",
  "成了",
  "它的",
  "玩伴",
  "，",
  "它",
  "时而",
  "扑",
  "腾",
  "着",
  "追逐",
  "，",
  "时而",
  "安静",
  "地",
  "观察",
  "，",
  "一举",
  "一",
  "动",
  "都",
  "透露",
  "着",
  "童真",
  "与",
  "可爱",
  "。",
  "偶尔",
  "，",
  "它",
  "会",
  "停下",
  "来",
  "，",
  "昂首",
  "望向",
  "天空",
  "，",
  "仿佛",
  "在",
  "对",
  "着",
  "那",
  "洁白",
  "的",
  "云",
  "朵",
  "诉说着",
  "自己",
  "的",
  "小心",
  "事",
  "。",
  "阳光",
  "照耀",
  "在",
  "小猫",
  "柔软",
  "的",
  "毛发",
  "上",
  "，",
  "形成",
  "一层",
  "淡淡的",
  "光",
  "晕",
  "，",
  "这",
  "一刻",
  "，",
  "它",
  "就像",
  "是",
  "草丛",
  "中的",
  "小",
  "精灵",
  "，",
  "为",
  "这",
  "美好",
  "的",
  "天气",
  "增添",
  "了",
  "一抹",
  "生动",
  "的",
  "色彩",
  "。",
]
`;
